File: grooves-api/src/main/groovy/com/github/rahulsom/grooves/api/snapshots/VersionedSnapshot.java
package com.github.rahulsom.grooves.api.snapshots;

import com.github.rahulsom.grooves.api.AggregateType;
import com.github.rahulsom.grooves.api.events.BaseEvent;
import com.github.rahulsom.grooves.api.snapshots.internal.BaseSnapshot;

/**
 * Marks a class as a versioned snapshot.
 *
 * @param <AggregateT>  The Aggregate this snapshot works over
 * @param <SnapshotIdT> The type for the snapshot's {@link #getId()} field
 * @param <EventIdT>    The type for the {@link EventT}'s id field
 * @param <EventT>      The base type for events that apply to {@link AggregateT}
 *
 * @author Rahul Somasunderam
 */
public interface VersionedSnapshot<
        AggregateT extends AggregateType,
        SnapshotIdT,
        EventIdT,
        EventT extends BaseEvent<AggregateT, EventIdT, EventT>>
        extends BaseSnapshot<AggregateT, SnapshotIdT, EventIdT, EventT> {

    Long getLastEventPosition();

    void setLastEventPosition(Long id);

    @Override
    default void setLastEvent(EventT aidBaseEvent) {
        this.setLastEventPosition(aidBaseEvent.getPosition());
    }
}

File: grooves-types/src/main/kotlin/com/github/rahulsom/grooves/api/snapshots/VersionedSnapshot.kt
package com.github.rahulsom.grooves.api.snapshots

import com.github.rahulsom.grooves.api.events.BaseEvent
import com.github.rahulsom.grooves.api.snapshots.internal.BaseSnapshot

/**
 * Marks a class as a versioned snapshot.
 *
 * @param [AggregateT] The Aggregate this snapshot works over
 * @param [SnapshotIdT] The type for [BaseSnapshot.id]
 * @param [EventIdT] The type for [BaseEvent.id]
 * @param [EventT] The base type for events that apply to [AggregateT]
 *
 * @author Rahul Somasunderam
 */
interface VersionedSnapshot<
        AggregateT,
        SnapshotIdT,
        EventIdT,
        in EventT : BaseEvent<AggregateT, EventIdT, in EventT>> :
    BaseSnapshot<AggregateT, SnapshotIdT, EventIdT, EventT> {

    /**
     * The position of the last event that this snapshot represents.
     *
     * This is useful in finding a matching or suitable snapshot based on the version of a snapshot.
     */
    var lastEventPosition: Long
}